Convert JavaScriptCore to Modern Objective-C

LumioseSil opened the pull request:
#### 5828788aa36b103f16a1990848521e05c1100c00
<pre>
Convert JavaScriptCore to Modern Objective-C
<a href="https://bugs.webkit.org/show_bug.cgi?id=251797">https://bugs.webkit.org/show_bug.cgi?id=251797</a>

Reviewed by NOBODY (OOPS!).

JavaScriptCore was run through the Xcode&apos;s Modern Objective-C converter.
However, the results were not buildable alone, so some additional
changes were made to get the JavaScriptCore framework to build again.

As this is an API change, this warrants modernization as:
1. This code has roots going back years, which makes sense as Safari has
   been around for 2 decades. Modernizing the code helps as it lays the
   groundwork for future patches to follow modern Objective-C coding
   conventions as well as being consistent with newer code written.
2. These modernizations make the code more safe while improving
   performance in some places.
3. If we ever need to write Swift code calling this library directly, it
   will be much easier to get Swift to work with this framework.

* Source/JavaScriptCore/API/APICallbackFunction.h:
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
* Source/JavaScriptCore/API/JSContext.mm:
* Source/JavaScriptCore/API/JSContextInternal.h:
* Source/JavaScriptCore/API/JSManagedValue.h:
* Source/JavaScriptCore/API/JSManagedValue.mm:
* Source/JavaScriptCore/API/JSScript.h:
* Source/JavaScriptCore/API/JSScript.mm:
* Source/JavaScriptCore/API/JSScriptInternal.h:
* Source/JavaScriptCore/API/JSScriptSourceProvider.mm:
* Source/JavaScriptCore/API/JSValue.mm:
* Source/JavaScriptCore/API/JSVirtualMachine.mm:
* Source/JavaScriptCore/API/JSWrapperMap.mm:
* Source/JavaScriptCore/API/ObjCCallbackFunction.mm:
* Source/JavaScriptCore/API/tests/CurrentThisInsideBlockGetterTest.mm:
* Source/JavaScriptCore/API/tests/DateTests.mm:
* Source/JavaScriptCore/API/tests/Regress141275.mm:
* Source/JavaScriptCore/API/tests/Regress141809.mm:
* Source/JavaScriptCore/API/tests/testapi.mm:
* Source/JavaScriptCore/inspector/ContentSearchUtilities.cpp:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorXPCConnection.mm:
* Source/JavaScriptCore/testmem/testmem.mm:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
* Source/WTF/wtf/cocoa/LanguageCocoa.mm:
* Source/WTF/wtf/cocoa/NSURLExtras.mm:
* Source/WebCore/bridge/objc/objc_utility.mm:
* Source/WebCore/platform/mac/PluginBlocklist.mm:
* Source/WebGPU/WebGPU/Adapter.mm:
* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
* Source/WebKit/UIProcess/API/Cocoa/APIContentRuleListStoreCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfiguration.mm:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
* Source/bmalloc/bmalloc/ProcessCheck.mm:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f1fd02b36d7c19816ce093c3594498cd73e2e64

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/682 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/701 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/729 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/874 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/600 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/673 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/750 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/786 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/835 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/689 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/651 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/650 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/848 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/721 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/645 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/666 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/599 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/613 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/666 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/674 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/645 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/635 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/705 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/598 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/651 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/147 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/655 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/706 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/141 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -343,23 +349,9 @@ @implementation JSContext (Internal)
-
     JSC::JSGlobalObject* globalObject = toJS(context);
-    m_virtualMachine = [JSVirtualMachine virtualMachineWithContextGroupRef:toRef(&globalObject->vm())];
-    ASSERT(m_virtualMachine);
```
This code is not the same, right? Why are we removing these assertions?

AZero13 replied:
Fixed!

aproskuryakov commented on Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
```diff
@@ -222,10 +222,10 @@
     if (context.exception) {
         scope.release();
-        promise->reject(globalObject, toJS(globalObject, [context.exception JSValueRef]));
+        promise->reject(globalObject, toJS(globalObject, (context.exception).JSValueRef));
```
The parentheses seem unneeded?

aproskuryakov commented on Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
```diff
@@ -279,7 +279,7 @@
     auto scope = DECLARE_THROW_SCOPE(vm);
 
-    Identifier key = Identifier::fromString(vm, String { [[script sourceURL] absoluteString] });
+    Identifier key = Identifier::fromString(vm, String { [script sourceURL].absoluteString });
```
I just briefly scrolled through the PR, and do not intend on doing a full review, but the machine output definitely looks like it needs another cleanup pass.

aproskuryakov commented on Source/JavaScriptCore/API/JSContext.h:
```diff
@@ -57,6 +57,15 @@ JSC_CLASS_AVAILABLE(macos(10.9), ios(7.0))
+@param context JSGlobalContextRef the context will hold.
+@result The new context.
+*/
+- (instancetype)initWithVirtualMachine:(JSVirtualMachine *)virtualMachine context:(JSGlobalContextRef)context NS_DESIGNATED_INITIALIZER NS_UNAVAILABLE;
```
This adds a public API, which is a very significant change. It would block the PR on going through certain Apple internal review first.

AZero13 replied:
Xcode refuses to compile this any other way but I will see if I could find another solution

kmiller68 replied:
I think this method should be declared in JSContextInternal.h. That header is not exposed to clients of JSC.

AZero13 replied:
Tried that. Xcode will not build.

kmiller68 replied:
Yeah, the implementation needs to be inside the `@implementation JSContext (Internal)` block.

AZero13 replied:
definition or just the declaration or both?

kmiller68 replied:
If the declaration is inside the `@interface JSContext(Internal)` then the implementation needs to be inside the `@implementation JSContext (Internal)`. At least if I remember correctly.

AZero13 replied:
Not working sadly. Only the way I had it before works.

AZero13 replied:
@kmiller68 So we are at an impasse. Making it public is the only way this even builds

mcatanzaro replied:
You're trying to add new public API to fix... a build failure? That did not exist before your pull request?

🤔️

AZero13 replied:
Yeah. I know...

AZero13 replied:
> You're trying to add new public API to fix... a build failure? That did not exist before your pull request?
> 
> 🤔️

Yeah just read the build-logs.

AZero13 replied:
I fixed it. @mcatanzaro

JosephPecoraro commented on Source/JavaScriptCore/API/JSScript.h:
```diff
@@ -84,22 +84,19 @@ JSC_CLASS_AVAILABLE(macos(10.15), ios(13.0))
  @abstract Returns the sourceURL of this JSScript.
  */
-- (NSURL *)sourceURL;
+@property (nonatomic, readonly, copy) NSURL *_Nonnull sourceURL;
```
This is already inside of a `NS_ASSUME_NONNULL_BEGIN`/`END` block. Is the `_Nonnull` really needed here?

JosephPecoraro commented on Source/JavaScriptCore/API/JSManagedValue.mm:
```diff
@@ -85,7 +85,7 @@ - (instancetype)initWithValue:(JSValue *)value
         return self;
 
-    JSC::JSGlobalObject* globalObject = toJS([value.context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJS((value.context).JSGlobalContextRef);
```
Isn't JSGlobalObject* still a C++ type? So the `*` could stay where it was?
Also it seems the parenthesis may not be needed on the right side. Yielding `value.context.JSGlobalContextRef`.

JosephPecoraro commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -329,39 +315,45 @@ @implementation JSContext(SubscriptSupport)
     return self;
 }
 
+
```
Looks like an extra newline.

JosephPecoraro commented on Source/JavaScriptCore/API/JSScriptInternal.h:
```diff
@@ -44,11 +44,11 @@ class String;
+@property (nonatomic, readonly) unsigned hash;
+@property (nonatomic, readonly) const WTF::String &source;
+@property (nonatomic, readonly) RefPtr<JSC::CachedBytecode> cachedBytecode;
+@property (nonatomic, readonly) JSC::JSSourceCode *_Nonnull jsSourceCode;
```
Same comments as before around `&`/`*` placement and `_Nonnull` usage inside `NS_ASSUME_NONNULL_BEGIN`/`END` blocks.

JosephPecoraro commented on Source/bmalloc/bmalloc/ProcessCheck.mm:
```diff
@@ -37,13 +37,13 @@ bool gigacageEnabledForProcess()
     // If we wanted to make it efficient to call more than once, we could memoize the result in a global boolean.
 
-    NSString *appName = [[NSBundle mainBundle] bundleIdentifier];
+    NSString *appName = [NSBundle mainBundle].bundleIdentifier;
```
`NSBundle.mainBundle.bundleIdentifier`?

AZero13 replied:
NSBundle is the class name, not a variable name

JosephPecoraro commented on Source/bmalloc/bmalloc/ProcessCheck.mm:
```diff
@@ -37,13 +37,13 @@ bool gigacageEnabledForProcess()
     }
 
-    NSString *processName = [[NSProcessInfo processInfo] processName];
+    NSString *processName = [NSProcessInfo processInfo].processName;
```
`NSProcessInfo.processInfo.processName`?

Constellation commented on Source/JavaScriptCore/API/JSManagedValue.mm:
```diff
@@ -85,7 +85,7 @@ - (instancetype)initWithValue:(JSValue *)value
         return self;
 
-    JSC::JSGlobalObject* globalObject = toJS([value.context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJS(value.context.JSGlobalContextRef);
```
This is not correct. Our convention is `*` on left hand side for non Objective-C class. And JSGlobalObject* is not Objective-C class.

Constellation commented on Source/JavaScriptCore/API/JSContext.mm:
```diff
@@ -162,7 +162,7 @@ - (JSValue *)dependencyIdentifiersForModuleJSScript:(JSScript *)script
 
     auto scope = DECLARE_CATCH_SCOPE(vm);
-    JSC::JSArray* result = globalObject->moduleLoader()->dependencyKeysIfEvaluated(globalObject, JSC::jsString(vm, String([[script sourceURL] absoluteString])));
+    JSC::JSArray *result = globalObject->moduleLoader()->dependencyKeysIfEvaluated(globalObject, JSC::jsString(vm, String([script sourceURL].absoluteString)));
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSScript.mm:
```diff
@@ -197,7 +197,7 @@ - (void)readCache
     Ref<JSC::CachedBytecode> cachedBytecode = JSC::CachedBytecode::create(WTFMove(mappedFile));
 
-    JSC::VM& vm = *toJS([m_virtualMachine JSContextGroupRef]);
+    JSC::VM &vm = *toJS([m_virtualMachine JSContextGroupRef]);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSScript.mm:
```diff
@@ -265,11 +265,11 @@ - (unsigned)hash
 - (JSC::SourceCode)sourceCode
 {
-    JSC::VM& vm = *toJS([m_virtualMachine JSContextGroupRef]);
+    JSC::VM &vm = *toJS([m_virtualMachine JSContextGroupRef]);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSScript.mm:
```diff
@@ -280,7 +280,7 @@ - (unsigned)hash
 - (JSC::JSSourceCode*)jsSourceCode
 {
-    JSC::VM& vm = *toJS([m_virtualMachine JSContextGroupRef]);
+    JSC::VM &vm = *toJS([m_virtualMachine JSContextGroupRef]);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSScript.mm:
```diff
@@ -327,7 +327,7 @@ - (BOOL)writeCache:(String&)error
     JSC::BytecodeCacheError cacheError;
     JSC::SourceCode sourceCode = [self sourceCode];
-    JSC::VM& vm = *toJS([m_virtualMachine JSContextGroupRef]);
+    JSC::VM &vm = *toJS([m_virtualMachine JSContextGroupRef]);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSScriptInternal.h:
```diff
@@ -44,11 +44,11 @@ class String;
-- (JSC::JSSourceCode*)jsSourceCode;
-- (JSC::SourceCode)sourceCode;
+@property (nonatomic, readonly) unsigned hash;
+@property (nonatomic, readonly) const WTF::String &source;
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSScriptInternal.h:
```diff
@@ -44,11 +44,11 @@ class String;
+@property (nonatomic, readonly) unsigned hash;
+@property (nonatomic, readonly) const WTF::String &source;
+@property (nonatomic, readonly) RefPtr<JSC::CachedBytecode> cachedBytecode;
+@property (nonatomic, readonly) JSC::JSSourceCode *jsSourceCode;
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSValue.mm:
```diff
@@ -504,21 +504,21 @@ - (BOOL)isInstanceOf:(id)value
 - (JSValue *)callWithArguments:(NSArray *)argumentArray
 {
-    JSC::JSGlobalObject* globalObject = toJS([_context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJS(_context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSValue.mm:
```diff
@@ -527,21 +527,21 @@ - (JSValue *)callWithArguments:(NSArray *)argumentArray
 - (JSValue *)constructWithArguments:(NSArray *)argumentArray
 {
-    JSC::JSGlobalObject* globalObject = toJS([_context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJS(_context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSValue.mm:
```diff
@@ -550,30 +550,30 @@ - (JSValue *)constructWithArguments:(NSArray *)argumentArray
 - (JSValue *)invokeMethod:(NSString *)method withArguments:(NSArray *)arguments
 {
-    JSC::JSGlobalObject* globalObject = toJS([_context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJS(_context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -137,9 +137,9 @@ static bool constructorHasInstance(JSContextRef ctx, JSObjectRef constructorRef,
+    definition.className = brand.UTF8String;
     JSClassRef classRef = JSClassCreate(&definition);
-    JSC::JSObject* result = makeWrapper([context JSGlobalContextRef], classRef, cls);
+    JSC::JSObject *result = makeWrapper(context.JSGlobalContextRef, classRef, cls);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -148,10 +148,10 @@ static bool constructorHasInstance(JSContextRef ctx, JSObjectRef constructorRef,
     definition.hasInstance = constructorHasInstance;
     JSClassRef classRef = JSClassCreate(&definition);
-    JSC::JSObject* result = makeWrapper([context JSGlobalContextRef], classRef, cls);
+    JSC::JSObject *result = makeWrapper(context.JSGlobalContextRef, classRef, cls);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -179,28 +179,28 @@ static bool constructorHasInstance(JSContextRef ctx, JSObjectRef constructorRef,
+    if (!base.isObject)
         return;
-    JSC::JSGlobalObject* globalObject = toJS([context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJS(context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -179,28 +179,28 @@ static bool constructorHasInstance(JSContextRef ctx, JSObjectRef constructorRef,
     auto scope = DECLARE_CATCH_SCOPE(vm);
 
-    JSC::JSObject* baseObject = JSC::asObject(toJS(globalObject, [base JSValueRef]));
+    JSC::JSObject *baseObject = JSC::asObject(toJS(globalObject, base.JSValueRef));
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -266,7 +266,7 @@ static void copyMethodsToObject(JSContext *context, Class objcClass, Protocol *p
             if (!name)
                 name = selectorToPropertyName(nameCStr);
-            JSC::JSGlobalObject* globalObject = toJS([context JSGlobalContextRef]);
+            JSC::JSGlobalObject *globalObject = toJS(context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -489,7 +489,7 @@ - (ConstructorPrototypePair)allocateConstructorAndPrototypeInContext:(JSContext
 
     if (!superClassInfo) {
-        JSC::JSGlobalObject* globalObject = toJSGlobalObject([context JSGlobalContextRef]);
+        JSC::JSGlobalObject *globalObject = toJSGlobalObject(context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -543,7 +543,7 @@ - (ConstructorPrototypePair)allocateConstructorAndPrototypeInContext:(JSContext
     JSC::Structure* structure = [self structureInContext:context];
 
-    JSC::JSGlobalObject* globalObject = toJS([context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJS(context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -576,7 +576,7 @@ - (ConstructorPrototypePair)allocateConstructorAndPrototypeInContext:(JSContext
         return structure;
 
-    JSC::JSGlobalObject* globalObject = toJSGlobalObject([context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJSGlobalObject(context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/JavaScriptCore/API/ObjCCallbackFunction.mm:
```diff
@@ -708,7 +708,7 @@ static JSObjectRef objCCallbackFunctionForInvocation(JSContext *context, NSInvoc
     }
 
-    JSC::JSGlobalObject* globalObject = toJS([context JSGlobalContextRef]);
+    JSC::JSGlobalObject *globalObject = toJS(context.JSGlobalContextRef);
```
Ditto.

Constellation commented on Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
```diff
@@ -129,14 +129,14 @@ String openTemporaryFile(StringView prefix, PlatformFileHandle& platformFileHand
 
     NSString *tempDirectoryComponent = [directoryPrefix stringByAppendingString:@"-XXXXXXXX"];
-    const char* tempDirectoryCString = [[tempDirectory stringByAppendingPathComponent:tempDirectoryComponent] fileSystemRepresentation];
+    const char *tempDirectoryCString = [tempDirectory stringByAppendingPathComponent:tempDirectoryComponent].fileSystemRepresentation;
```
Ditto.

Constellation commented on Source/WTF/wtf/cocoa/NSURLExtras.mm:
```diff
@@ -45,7 +45,7 @@ static BOOL readIDNAllowedScriptListFile(NSString *filename)
         return NO;
 
-    FILE *file = fopen([filename fileSystemRepresentation], "r");
+    FILE *file = fopen(filename.fileSystemRepresentation, "r");
```
Ditto.

mcatanzaro commented on Source/JavaScriptCore/API/JSScript.h:
```diff
@@ -84,22 +84,22 @@ JSC_CLASS_AVAILABLE(macos(10.15), ios(13.0))
 
 /*!
- @method
+ @property
  @abstract Returns true when evaluating this JSScript will use the bytecode cache. Returns false otherwise.
  */
-- (BOOL)isUsingBytecodeCache;
+@property (nonatomic, getter=isUsingBytecodeCache, readonly) BOOL usingBytecodeCache;
```
I can't do much more here since I don't know Objective C, but there are still a lot of API changes in this pull request. Looks like they might possibly be backwards-compatible and therefore OK? Not sure. It's going to require Apple API review regardless.

AZero13 replied:
All backwards compatible.

AZero13 replied:
> I can't do much more here since I don't know Objective C, but there are still a lot of API changes in this pull request. Looks like they might possibly be backwards-compatible and therefore OK? Not sure. It's going to require Apple API review regardless.

Do you know who can do this?

xeenon commented on Source/JavaScriptCore/API/JSWrapperMap.mm:
```diff
@@ -275,7 +275,7 @@ static void copyMethodsToObject(JSContext *context, Class objcClass, Protocol *p
             // the existing value on the prototype chain is an ObjC
             // callback already.
-            if ([existingMethod isObject] && JSC::jsDynamicCast<JSC::ObjCCallbackFunction*>(toJS(globalObject, [existingMethod JSValueRef])))
+            if (existingMethod.isObject && JSC::jsDynamicCast<JSC::ObjCCallbackFunction *>(toJS(globalObject, existingMethod.JSValueRef)))
```
Revert the space between the type and `*` for C++ types.

xeenon commented on Source/WTF/wtf/cocoa/NSURLExtras.mm:
```diff
@@ -140,13 +140,13 @@ static String decodePercentEscapes(const String& string)
         baseURL = URLByTruncatingOneCharacterBeforeComponent(baseURL, kCFURLComponentResourceSpecifier);
 
-        const UInt8 *bytes = static_cast<const UInt8*>([data bytes]);
+        const UInt8 *bytes = static_cast<const UInt8 *>(data.bytes);
```
Remove space before `*`.

xeenon commented on Source/WTF/wtf/cocoa/NSURLExtras.mm:
```diff
@@ -317,7 +317,7 @@ static bool hasQuestionMarkOnlyQueryString(NSURL *URL)
 {
     NSData *data = originalURLData(URL);
-    return URLHelpers::userVisibleURL(CString(static_cast<const char*>([data bytes]), [data length]));
+    return URLHelpers::userVisibleURL(CString(static_cast<const char *>(data.bytes), data.length));
```
Remove space before `*`.

emw-apple commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -234,13 +234,13 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
 @result The Objective-C representation of this JSValue.
 */
-- (id)toObject;
+@property (nonatomic, readonly, strong) id toObject;
```
Here's an example of something I think the migrator does wrong. This is a _conversion_, not an accessor. I don't think it's appropriate to expose it as a getter property, because that hides that the implementation is doing nontrivial work. i.e. in Swift, it _should_ be `JSValue.toObject()`, not `getter:JSValue.toObject`.

AZero13 replied:
Fixed!

emw-apple reviewed (commented):
I understand that this PR is mostly mechanical, but as the past cycles of review have shown, it's not something that we can rubber-stamp without detailed analysis, and the Objective-C migrator does not necessarily agree with our API design decisions.

Re-posting @aproskuryakov's comment, since it's now buried in EWS notifications:

> I am sorry that you are spending effort constantly rebasing. It is not anyone's intention to waste your time.
>
> Changing public API needs to be done separately from refactoring the implementation, and it needs an extremely strong rationale of its own. Please explain why these API changes are necessary for developers, in a way that's strong and obvious enough to warrant time investment from many more people.

Addressing the points in your commit message:

> As this is an API change, this warrants modernization as:
>
> 1. This code has roots going back years, which makes sense as Safari has
>    been around for 2 decades. Modernizing the code helps as it lays the
>    groundwork for future patches to follow modern Objective-C coding
>    conventions as well as being consistent with newer code written.

This is a good intuition! But it's hard to change so many small details at once. Each declaration in JSC's API needs heightened scrutiny and consensus.

> 2. These modernizations make the code more safe while improving
>    performance in some places.

Please call out the safety and performance improvements. From glancing at the diff, it looks like churn. We should be extra aware of any functional changes in the implementation.

> 3. If we ever need to write Swift code calling this library directly, it
>    will be much easier to get Swift to work with this framework.

JavaScriptCore is already usable from Swift. Changing the API to be more ergonomic in Swift will require a lot of review from Apple contributors (i.e. more than just my or someone else's r+).

I defer to @WebKit/jsc-reviewers, but it's hard for me to see a path forward for this PR. I recommend you talk to us on [the WebKit Slack](https://webkit.org/getting-started/#staying-in-touch) so that we can agree about what (if any) Objective-C modernizations we want to take. I don't want to waste your time rebasing without clarity about what JSC will accept.

LumioseSil committed:
Convert JavaScriptCore to Modern Objective-C
https://bugs.webkit.org/show_bug.cgi?id=251797

Reviewed by NOBODY (OOPS!).

JavaScriptCore was run through the Xcode's Modern Objective-C converter.
However, the results were not buildable alone, so some additional
changes were made to get the JavaScriptCore framework to build again.

As this is an API change, this warrants modernization as:
1. This code has roots going back years, which makes sense as Safari has
   been around for 2 decades. Modernizing the code helps as it lays the
   groundwork for future patches to follow modern Objective-C coding
   conventions as well as being consistent with newer code written.
2. These modernizations make the code more safe while improving
   performance in some places.
3. If we ever need to write Swift code calling this library directly, it
   will be much easier to get Swift to work with this framework.

* Source/JavaScriptCore/API/APICallbackFunction.h:
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
* Source/JavaScriptCore/API/JSContext.mm:
* Source/JavaScriptCore/API/JSContextInternal.h:
* Source/JavaScriptCore/API/JSManagedValue.h:
* Source/JavaScriptCore/API/JSManagedValue.mm:
* Source/JavaScriptCore/API/JSScript.h:
* Source/JavaScriptCore/API/JSScript.mm:
* Source/JavaScriptCore/API/JSScriptInternal.h:
* Source/JavaScriptCore/API/JSScriptSourceProvider.mm:
* Source/JavaScriptCore/API/JSValue.mm:
* Source/JavaScriptCore/API/JSVirtualMachine.mm:
* Source/JavaScriptCore/API/JSWrapperMap.mm:
* Source/JavaScriptCore/API/ObjCCallbackFunction.mm:
* Source/JavaScriptCore/API/tests/CurrentThisInsideBlockGetterTest.mm:
* Source/JavaScriptCore/API/tests/DateTests.mm:
* Source/JavaScriptCore/API/tests/Regress141275.mm:
* Source/JavaScriptCore/API/tests/Regress141809.mm:
* Source/JavaScriptCore/API/tests/testapi.mm:
* Source/JavaScriptCore/inspector/ContentSearchUtilities.cpp:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorXPCConnection.mm:
* Source/JavaScriptCore/testmem/testmem.mm:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
* Source/WTF/wtf/cocoa/LanguageCocoa.mm:
* Source/WTF/wtf/cocoa/NSURLExtras.mm:
* Source/WebCore/bridge/objc/objc_utility.mm:
* Source/WebCore/platform/mac/PluginBlocklist.mm:
* Source/WebGPU/WebGPU/Adapter.mm:
* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
* Source/WebKit/UIProcess/API/Cocoa/APIContentRuleListStoreCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfiguration.mm:
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
* Source/bmalloc/bmalloc/ProcessCheck.mm:
